### PR TITLE
scorecard - add return code logic

### PR DESCRIPTION
**Pull request by jmccormick2001**
this PR adds return code logic into the scorecard subcommand, so that if any selected
test fails, a return code of 1 is produced.  If no tests fail, then the normal return code of 0 is produced.

**Description of the change:**
adds return code logic, updates the scorecard docs to mention the new logic, and also adjusts the scorecard tests to allow for non-zero return codes.

**Motivation for the change:**

this change is part of the scorecard v1alpha2 implementation and supports pipelines that run the scorecard in particular so that they can look for non-zero return codes instead of having to parse the scorecard output.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -53,6 +53,7 @@
 - Added new `--version` flag to the `operator-sdk scorecard` command to support a new output format for the scorecard. ([#1916](https://github.com/operator-framework/operator-sdk/pull/1916)
 - Added new `--selector` flag to the `operator-sdk scorecard` command to support filtering scorecard tests based on labels added to each test. ([#1916](https://github.com/operator-framework/operator-sdk/pull/1916)
 - Added new `--list` flag to the `operator-sdk scorecard` command to support listing scorecard tests that would be executed based on selector filters. ([#1916](https://github.com/operator-framework/operator-sdk/pull/1916)
+- For scorecard version v1alpha2 only, return code logic was added to return 1 if any of the selected scorecard tests fail.  A return code of 0 is returned if all selected tests pass. ([#1916](https://github.com/operator-framework/operator-sdk/pull/1916)
```
Comment (camilamacedo86) [inline CHANGELOG.md]:
```suggestion
- Change points result for pass/fail in the scorecard version v1alpha2 only. ([#1916](https://github.com/operator-framework/operator-sdk/pull/1916)
```
WDYT? something more user-friendly ;-).

File: doc/test-framework/scorecard.md [hunk 2]
```diff
@@ -49,6 +49,7 @@ Following are some requirements for the operator project which would be  checked
 1. Setup the `.osdk-scorecard.yaml` configuration file in your project.. See [Config file](#config-file)
 2. Create the namespace defined in the RBAC files(`role_bindinding`)
 3. Then, run the scorecard, for example `$ operator-sdk scorecard`. See the [Command args](#command-args) to check its options.
+4. Note, starting with version v1alpha2, the scorecard return code is 1 if any of the tests executed did not pass and 0 if all selected tests pass.
```
Comment (camilamacedo86) [inline doc/test-framework/scorecard.md]: I think it is not required here.

Comment (camilamacedo86) [reply doc/test-framework/scorecard.md]: Maybe add some NOTE: as if you use version 2 then the output will be 0 or 1 to indicate that the test pass or fails respectively.  But, IMHO not here. It could be done where has the detailed info over its results/outputs.

File: hack/tests/subcommand-scorecard.sh [hunk 3]
```diff
@@ -16,10 +16,13 @@ set -ex
 pushd test/test-framework
 
 # test to see if scorecard fails when version is v1alpha2 and when external plugins are configured
-operator-sdk scorecard --version v1alpha2 --config "$CONFIG_PATH" |& grep '^.*error validating plugin config.*$'
+!(operator-sdk scorecard --version v1alpha2 --config "$CONFIG_PATH" |& grep '^.*error validating plugin config.*$')
```
Comment (camilamacedo86) [inline hack/tests/subcommand-scorecard.sh]: Why ! was added?

Comment (joelanford) [reply hack/tests/subcommand-scorecard.sh]: To make things easier to debug when this fails, it may be more clear to do something like:
```bash
if ! $(operator-sdk scorecard --version v1alpha2 --config "$CONFIG_PATH" |& grep -q '^.*error validating plugin config.*$'); then
    echo "expected scorecard to fail when version is v1alpha2 and when external plugins are configured"
    exit 1
fi

File: hack/tests/subcommand-scorecard.sh [hunk 4]
```diff
@@ -48,6 +51,8 @@ then
 	exit 1
 fi
 
+set -xe
```
Comment (joelanford) [inline hack/tests/subcommand-scorecard.sh]: Is there a way to refactor this such that we can leave `set -xe` enabled even while expecting this command to exit non-zero? Maybe something similar to the `if` statement in my other comment?

That way if some command fails that we don't expect to fail, then we'll still fail the test.

File: CHANGELOG.md (hunk 1 above)
Comment (jmccormick2001) [reply CHANGELOG.md]: I like the current comment better I think, this change is about the exit code that scorecard returns instead of a change to the points.

File: doc/test-framework/scorecard.md (hunk 2 above)
Comment (jmccormick2001) [reply doc/test-framework/scorecard.md]: I could make it a separate section like man pages do, something like a heading of EXIT STATUS if that makes it clearer.  My intent here is to let the end user know what the exit codes will be out of scorecard so I figured this was the correct document for that communication.

Comment (camilamacedo86) [reply doc/test-framework/scorecard.md]: It is fine. But this section is describing the 3 macro steps to set up the scorecard, so really make sense the 4 step is a description of the result when the v2 is used?

File: internal/scorecard/scorecard_output.go [hunk 5]
```diff
@@ -43,26 +43,28 @@ func printPluginOutputs(version string, pluginOutputs []scapiv1alpha1.ScorecardO
...
-		return nil
+	}
 
+	if schelpers.IsV1alpha2(version) {
+		testOutput := list.(scapiv1alpha2.ScorecardOutput)
+		for _, result := range testOutput.Results {
+			if result.State == scapiv1alpha2.FailState {
+				return fmt.Errorf("test %s, did not pass", result.Name)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (jmccormick2001) [reply internal/scorecard/scorecard_output.go]: https://jira.coreos.com/browse/OSDK-586

a story for that configuration refactor is above.

File: doc/test-framework/scorecard.md (hunk 2 above)
Comment (jmccormick2001) [reply doc/test-framework/scorecard.md]: I added an Exit Status heading to call this out separately.

File: doc/test-framework/scorecard.md [hunk 6]
```diff
@@ -52,6 +52,10 @@ Following are some requirements for the operator project which would be  checked
 
 **NOTE:** If your operator is non-SDK then some steps will be required in order to meet its requirements.
 
+## Exit Status
```
Comment (camilamacedo86) [inline doc/test-framework/scorecard.md]: Note that the doc follow a sequence of information which makes easier the understanding. So, in this way, I do not think that this section would be helpful here.  WDYT about add it after https://github.com/operator-framework/operator-sdk/blob/f0f61013dea6702466d747b9a3141f2275a5f239/doc/test-framework/scorecard.md#tests-performed? Also, o not forget to update the index.

File: internal/scorecard/scorecard.go [hunk 7]
```diff
@@ -140,10 +141,21 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
...
 
-	if err := printPluginOutputs(scViper.GetString(schelpers.VersionOpt), pluginOutputs); err != nil {
+	output, err := printPluginOutputs(scViper.GetString(schelpers.VersionOpt), pluginOutputs)
+	if err != nil {
 		return err
 	}
 
+	apiVersion := scViper.GetString(schelpers.VersionOpt)
+	if schelpers.IsV1alpha2(apiVersion) {
+		testOutput := output.(scapiv1alpha2.ScorecardOutput)
```
Comment (joelanford) [inline internal/scorecard/scorecard.go]: This might be getting into out-of-scope refactor territory, but it's a little odd to me that we would need the result of a function that prints output to determine the state of our test run.

Is it possible to use `pluginOutputs` here and keep the signature of `printPluginOutputs()` as is?

File: doc/test-framework/scorecard.md [hunk 8]
```diff
@@ -177,6 +174,10 @@ Following the description of each internal [Plugin](#plugins). Note that are 8 i
 | Spec Fields With Descriptors | This test verifies that every field in the Custom Resources' spec sections have a corresponding descriptor listed in the CSV. This test has a maximum score equal to the total number of fields in the spec sections of each custom resource passed in via the `cr-manifest` option. |
 | Status Fields With Descriptors | This test verifies that every field in the Custom Resources' status sections have a corresponding descriptor listed in the CSV. This test has a maximum score equal to the total number of fields in the status sections of each custom resource passed in via the `cr-manifest` option. |
 
+## Exit Status
+
+In version v1alpha2, the scorecard return code is 1 if any of the tests executed did not pass and 0 if all selected tests pass.
```
Comment (camilamacedo86) [inline doc/test-framework/scorecard.md]: IHMO it still not fit in the doc. Note that if you add a section to explain the exit status you need say if you use v1alpha1 then .. however, if you are calling the scorecard with v1alpha2 then ...

File: hack/tests/subcommand-scorecard.sh [hunk 9]
```diff
@@ -16,36 +16,43 @@ set -ex
...
 
 # test to see if v1alpha2 is used from the command line
-commandoutput="$(operator-sdk scorecard --version v1alpha2 --config "$CONFIG_PATH_V1ALPHA2" 2>&1)"
-failCount=`echo $commandoutput | grep -o "fail" | wc -l`
-expectedFailCount=3
-if [ $failCount -ne $expectedFailCount ]
-then
-	echo "expected fail count $expectedFailCount, got $failCount"
-	exit 1
+if ! commandoutput="$(operator-sdk scorecard --version v1alpha2 --config "$CONFIG_PATH_V1ALPHA2" 2>&1)"; then
```
Comment (joelanford) [inline hack/tests/subcommand-scorecard.sh]: If this command exits with `0` exit code, we'll skip this block of tests. Should we fail if it exits `0`?

Similar question for the below tests as well. Should we have logic that handles the case where the command returns an unexpected exit code?

File: hack/tests/subcommand-scorecard.sh [hunk 10]
```diff
@@ -16,36 +16,43 @@ set -ex
...
 
 # test to see if list flag work
-commandoutput="$(operator-sdk scorecard --version v1alpha2 --list --selector=suite=basic --config "$CONFIG_PATH_V1ALPHA2" 2>&1)"
-labelCount=`echo $commandoutput | grep -o "Label" | wc -l`
-expectedLabelCount=3
-if [ $labelCount -ne $expectedLabelCount ]
-then
-	echo "expected label count $expectedLabelCount, got $labelCount"
-	exit 1
+if ! commandoutput="$(operator-sdk scorecard --version v1alpha2 --list --selector=suite=basic --config "$CONFIG_PATH_V1ALPHA2" 2>&1)"; then
```
Comment (joelanford) [inline hack/tests/subcommand-scorecard.sh]: Should this invocation with `--list` return 0? Looks like we're expecting it to return non-zero?

File: doc/test-framework/scorecard.md (hunk 8 above)
Comment (jmccormick2001) [reply doc/test-framework/scorecard.md]: updated to include a comment about what happens in the v1alpha1 case.

File: hack/tests/subcommand-scorecard.sh (hunk 9 above)
Comment (jmccormick2001) [reply hack/tests/subcommand-scorecard.sh]: should be fixed with this last update, the tests now only use the bash inversion (!) when the expected return code is 1, I added an 'else' to those cases if a 1 is not received. 

when the test is expected to return a 0, I let the 'set -e' do its thing and don't use bash inversion.

File: hack/tests/subcommand-scorecard.sh (hunk 10 above)
Comment (jmccormick2001) [reply hack/tests/subcommand-scorecard.sh]: fixed in latest update.

File: doc/test-framework/scorecard.md [hunk 11]
```diff
@@ -173,6 +174,10 @@ Following the description of each internal [Plugin](#plugins). Note that are 8 i
 | Spec Fields With Descriptors | This test verifies that every field in the Custom Resources' spec sections have a corresponding descriptor listed in the CSV. This test has a maximum score equal to the total number of fields in the spec sections of each custom resource passed in via the `cr-manifest` option. |
 | Status Fields With Descriptors | This test verifies that every field in the Custom Resources' status sections have a corresponding descriptor listed in the CSV. This test has a maximum score equal to the total number of fields in the status sections of each custom resource passed in via the `cr-manifest` option. |
 
+## Exit Status
+
+In version v1alpha2, the scorecard return code is 1 if any of the tests executed did not pass and 0 if all selected tests pass.  Version v1alpha1 returns exit code of 0 regardless if tests fail.
+
```
Comment (camilamacedo86) [inline doc/test-framework/scorecard.md]:
```suggestion
The scorecard called with the flag `--version=v1alpha2` will return `0` when all selected tests are executed successfully and `1` if any of them fails. Note that if you do not inform this flag and use the version `v1alpha1` all tests will return exit code 0 regardless if any of the tests fail.
```
